File: X410_CombinedTxRx_XL - 副本 - 副本/plot_latest_log_comparison.m
% Plot latest algorithm comparison from newest log statistics

algorithms = {'AARF','OLLA','AOLLA','QLOLLA'};
scriptDir = fileparts(mfilename('fullpath'));
logDir = fullfile(scriptDir, 'logs');

throughput = zeros(1, numel(algorithms));
per = zeros(1, numel(algorithms));
pickedLogs = strings(1, numel(algorithms));

for k = 1:numel(algorithms)
    pattern = fullfile(logDir, sprintf('run_%s_*.log', algorithms{k}));
    files = dir(pattern);
    if isempty(files)
        error('No log found for %s under %s', algorithms{k}, logDir);
    end
    [~, idx] = max([files.datenum]);
    logFile = fullfile(files(idx).folder, files(idx).name);
    pickedLogs(k) = string(files(idx).name);

    txt = fileread(logFile);

    fb = regexp(txt, '^Iter\s+\d+\s+\|\s+fb=(ack|arq)\s+\|', 'tokens', 'lineanchors');
    ackCount = 0;
    arqCount = 0;
    for i = 1:numel(fb)
        tag = fb{i}{1};
        if strcmp(tag, 'ack')
            ackCount = ackCount + 1;
        elseif strcmp(tag, 'arq')
            arqCount = arqCount + 1;
        end
    end
    if ackCount + arqCount > 0
        per(k) = arqCount / (ackCount + arqCount);
    else
        per(k) = 0;
    end

    bitrateTokens = regexp(txt, '^\s+DataBitrate \(Mbps\):\s*([0-9.]+)', 'tokens', 'lineanchors');
    bitrateVals = zeros(1, numel(bitrateTokens));
    for i = 1:numel(bitrateTokens)
        bitrateVals(i) = str2double(bitrateTokens{i}{1});
    end
    if ~isempty(bitrateVals)
        throughput(k) = mean(bitrateVals);
    else
        throughput(k) = 0;
    end
end

fig = figure('Color','w','Position',[100 100 1100 620]);
ax1 = axes(fig);
hold(ax1,'on');

b1 = bar(ax1, throughput, 0.62, ...
    'FaceColor',[0.36 0.61 0.84], ...
    'EdgeColor',[0.12 0.12 0.12], ...
    'LineWidth',0.8);

yyaxis right
b2 = bar(per, 0.62, ...
    'FaceColor',[0.93 0.49 0.19], ...
    'FaceAlpha',0.65, ...
    'EdgeColor',[0.12 0.12 0.12], ...
    'LineWidth',0.8);
ylabel('PER');

yyaxis left
ylabel('Throughput (Mbps)');
xlabel('Algorithm');
title('Latest Algorithm Comparison: Throughput vs PER');
set(ax1,'XTick',1:numel(algorithms),'XTickLabel',algorithms);
grid(ax1,'on');
ax1.GridAlpha = 0.2;

yyaxis left
ylim([0, max(throughput)*1.25 + 1]);
for i = 1:numel(throughput)
    text(i, throughput(i) + 0.5, sprintf('%.2f', throughput(i)), ...
        'HorizontalAlignment','center', 'Color',[0.18 0.46 0.71], 'FontSize',10);
end

yyaxis right
ylim([0, max(per)*1.15 + 0.02]);
for i = 1:numel(per)
    text(i, per(i) + 0.02, sprintf('%.2f', per(i)), ...
        'HorizontalAlignment','center', 'Color',[0.77 0.35 0.07], 'FontSize',10);
end

legend([b1 b2], {'Throughput','PER'}, 'Location','northwest');

outFile = fullfile(fileparts(mfilename('fullpath')), 'latest_log_comparison.png');
exportgraphics(fig, outFile, 'Resolution', 150);
fprintf('Saved plot to: %s\n', outFile);
for k = 1:numel(algorithms)
    fprintf('%s -> throughput=%.2f Mbps, PER=%.2f, log=%s\n', ...
        algorithms{k}, throughput(k), per(k), pickedLogs(k));
end

File: X410_CombinedTxRx_XL - 副本 - 副本/plot_latest_log_comparison.py
from pathlib import Path
import re
import statistics

import matplotlib.pyplot as plt
import numpy as np


ALGORITHMS = ["AARF", "OLLA", "AOLLA", "QLOLLA"]


def find_latest_log(log_dir: Path, algorithm: str) -> Path:
    matches = sorted(log_dir.glob(f"run_{algorithm}_*.log"))
    if not matches:
        raise FileNotFoundError(f"No log found for {algorithm} in {log_dir}")
    return matches[-1]


def parse_log(log_path: Path) -> dict:
    text = log_path.read_text(encoding="utf-8", errors="ignore")

    feedback = re.findall(r"^Iter\s+\d+\s+\|\s+fb=(ack|arq)\s+\|", text, re.M)
    ack = sum(1 for item in feedback if item == "ack")
    arq = sum(1 for item in feedback if item == "arq")
    per = arq / (ack + arq) if (ack + arq) else 0.0

    bitrate_vals = [
        float(x)
        for x in re.findall(r"^\s+DataBitrate \(Mbps\):\s*([0-9.]+)", text, re.M)
    ]
    avg_bitrate = statistics.mean(bitrate_vals) if bitrate_vals else 0.0

    final_match = re.search(
        r"^Final state \| MCS=(\d+) AMPDU=(\d+) BWdec=(\d+) lastSNR=([0-9.]+) dB",
        text,
        re.M,
    )
    stats_match = re.search(
        r"^Stats \| fallback=(\d+) \([^)]+\), realDecode=(\d+) \(([^)]+)\), qUpdate=(\d+)",
        text,
        re.M,
    )

    return {
        "ack": ack,
        "arq": arq,
        "per": per,
        "avg_bitrate_mbps": avg_bitrate,
        "final_mcs": int(final_match.group(1)) if final_match else -1,
        "last_snr_db": float(final_match.group(4)) if final_match else 0.0,
        "real_decode_pct": stats_match.group(3) if stats_match else "N/A",
    }


def collect_latest_results():
    log_dir = Path(__file__).with_name("logs")
    results = []
    for algorithm in ALGORITHMS:
        log_path = find_latest_log(log_dir, algorithm)
        metrics = parse_log(log_path)
        metrics["algorithm"] = algorithm
        metrics["log_file"] = log_path.name
        results.append(metrics)
    return results


def main():
    results = collect_latest_results()
    algorithms = [item["algorithm"] for item in results]
    throughput = [item["avg_bitrate_mbps"] for item in results]
    per = [item["per"] for item in results]

    x = np.arange(len(algorithms))
    width = 0.62

    fig, ax1 = plt.subplots(figsize=(10, 6), dpi=150)
    ax2 = ax1.twinx()

    bars1 = ax1.bar(
        x,
        throughput,
        width=width,
        color="#5B9BD5",
        edgecolor="#1F1F1F",
        linewidth=0.8,
        label="Throughput",
    )
    bars2 = ax2.bar(
        x,
        per,
        width=width,
        color="#ED7D31",
        alpha=0.65,
        edgecolor="#1F1F1F",
        linewidth=0.8,
        label="PER",
    )

    ax1.set_title("Latest Algorithm Comparison: Throughput vs PER")
    ax1.set_xlabel("Algorithm")
    ax1.set_ylabel("Throughput (Mbps)", color="#2F75B5")
    ax2.set_ylabel("PER", color="#C55A11")

    ax1.set_xticks(x)
    ax1.set_xticklabels(algorithms)
    ax1.grid(axis="y", linestyle="--", alpha=0.25)

    ax1.set_ylim(0, max(throughput) * 1.25 + 1 if throughput else 1)
    ax2.set_ylim(0, max(per) * 1.15 + 0.02 if per else 0.1)

    for bar, val in zip(bars1, throughput):
        ax1.text(
            bar.get_x() + bar.get_width() / 2,
            bar.get_height() + 0.35,
            f"{val:.2f}",
            ha="center",
            va="bottom",
            fontsize=9,
            color="#2F75B5",
        )

    for bar, val in zip(bars2, per):
        ax2.text(
            bar.get_x() + bar.get_width() / 2,
            bar.get_height() + 0.01,
            f"{val:.2f}",
            ha="center",
            va="bottom",
            fontsize=9,
            color="#C55A11",
        )

    fig.tight_layout()

    out_path = Path(__file__).with_name("latest_log_comparison.png")
    fig.savefig(out_path, bbox_inches="tight")
    print(f"Saved plot to: {out_path}")
    for item in results:
        print(
            f"{item['algorithm']}: "
            f"throughput={item['avg_bitrate_mbps']:.2f} Mbps, "
            f"PER={item['per']:.2f}, "
            f"log={item['log_file']}"
        )


if __name__ == "__main__":
    main()
